File: examples/http_client.js
#!/usr/bin/env node

/**
 * PM4Py REST API Client Examples (JavaScript/Node.js)
 *
 * Demonstrates how to use the PM4Py REST API from JavaScript/Node.js applications.
 * This example covers:
 * - Authentication
 * - Process discovery
 * - Conformance checking
 * - Log statistics
 * - Error handling and rate limiting
 *
 * Installation:
 *   npm install node-fetch
 *
 * Usage:
 *   export PM4PY_API_KEY="your-api-key"
 *   node examples/http_client.js
 *
 * Or with npm scripts in package.json:
 *   "scripts": { "api-examples": "node examples/http_client.js" }
 *   npm run api-examples
 */

const fetch = require("node-fetch");

/**
 * PM4Py REST API Client
 */
class PM4PyClient {
  constructor(apiKey, baseUrl = "http://localhost:8080/api/v1") {
    this.apiKey = apiKey;
    this.baseUrl = baseUrl.replace(/\/$/, ""); // Remove trailing slash
    this.rateLimitRemaining = 10000;
    this.rateLimitReset = null;
  }

  /**
   * Update rate limit information from response headers
   */
  updateRateLimitInfo(response) {
    const remaining = response.headers.get("X-RateLimit-Remaining");
    const reset = response.headers.get("X-RateLimit-Reset");

    if (remaining) {
      this.rateLimitRemaining = parseInt(remaining);
    }
    if (reset) {
      this.rateLimitReset = reset;
    }
  }

  /**
   * Make HTTP request to API
   */
  async makeRequest(method, path, data = null) {
    const url = `${this.baseUrl}${path}`;

    const options = {
      method,
      headers: {
        "X-API-Key": this.apiKey,
        "Content-Type": "application/json",
      },
    };

    if (data) {
      options.body = JSON.stringify(data);
    }

    try {
      const response = await fetch(url, options);

      // Update rate limit info
      this.updateRateLimitInfo(response);

      // Handle errors
      if (response.status === 429) {
        throw new Error(
          `Rate limit exceeded. Reset at ${this.rateLimitReset}`
        );
      }

      if (!response.ok) {
        const errorBody = await response.json().catch(() => ({}));
        throw new Error(
          `HTTP ${response.status}: ${errorBody.message || response.statusText}`
        );
      }

      return await response.json();
    } catch (error) {
      console.error(`API Error: ${error.message}`);
      throw error;
    }
  }

  /**
   * Check API health
   */
  async healthCheck() {
    return this.makeRequest("GET", "/health");
  }

  /**
   * Discover process model from event log
   *
   * @param {Array} events - List of events
   * @param {string} algorithm - Discovery algorithm (inductive, alpha, etc.)
   * @param {Object} parameters - Algorithm-specific parameters
   */
  async discover(events, algorithm = "inductive", parameters = {}) {
    const payload = {
      log: {
        events,
        format: "json",
      },
      algorithm,
    };

    if (Object.keys(parameters).length > 0) {
      payload.parameters = parameters;
    }

    return this.makeRequest("POST", "/discover", payload);
  }

  /**
   * Check log conformance against model
   *
   * @param {Array} events - List of events
   * @param {Object} model - Petri Net model
   * @param {string} variant - Conformance variant
   */
  async conform(events, model, variant = "token_replay") {
    const payload = {
      log: {
        events,
        format: "json",
      },
      model,
      variant,
    };

    return this.makeRequest("POST", "/conform", payload);
  }

  /**
   * Analyze model structure
   */
  async analyze(model) {
    return this.makeRequest("POST", "/analyze", { model });
  }

  /**
   * Extract log statistics
   */
  async stats(events) {
    const payload = {
      log: {
        events,
        format: "json",
      },
    };

    return this.makeRequest("POST", "/stats", payload);
  }
}

/**
 * Create sample event log for testing
 */
function createSampleEvents() {
  const events = [];
  const now = new Date();

  for (let caseNum = 1; caseNum <= 5; caseNum++) {
    const caseId = `loan_${String(caseNum).padStart(3, "0")}`;
    let caseTime = new Date(now);
    caseTime.setHours(caseTime.getHours() + caseNum);

    // Event 1: Apply
    events.push({
      case_id: caseId,
      activity: "apply",
      timestamp: caseTime.toISOString(),
      resource: "customer",
      attributes: { amount: 50000 * caseNum },
    });

    // Event 2: Register
    let time2 = new Date(caseTime);
    time2.setMinutes(time2.getMinutes() + 5);
    events.push({
      case_id: caseId,
      activity: "register",
      timestamp: time2.toISOString(),
      resource: "clerk",
    });

    // Event 3: Verify documents
    let time3 = new Date(caseTime);
    time3.setHours(time3.getHours() + 1);
    events.push({
      case_id: caseId,
      activity: "verify_documents",
      timestamp: time3.toISOString(),
      resource: "officer",
    });

    // Event 4: Credit check
    let time4 = new Date(caseTime);
    time4.setHours(time4.getHours() + 2);
    events.push({
      case_id: caseId,
      activity: "credit_check",
      timestamp: time4.toISOString(),
      resource: "system",
    });

    // Event 5: Approve
    let time5 = new Date(caseTime);
    time5.setHours(time5.getHours() + 3);
    events.push({
      case_id: caseId,
      activity: "approve",
      timestamp: time5.toISOString(),
      resource: "manager",
      attributes: { approved: true },
    });

    // Event 6: Disburse
    let time6 = new Date(caseTime);
    time6.setHours(time6.getHours() + 4);
    events.push({
      case_id: caseId,
      activity: "disburse",
      timestamp: time6.toISOString(),
      resource: "accountant",
      attributes: { amount_disbursed: 50000 * caseNum },
    });
  }

  return events;
}

/**
 * Format percentage
 */
function formatPercent(value) {
  return `${(value * 100).toFixed(1)}%`;
}

/**
 * Run examples
 */
async function main() {
  console.log("╔════════════════════════════════════════════════════════╗");
  console.log("║   PM4Py REST API Client Examples (JavaScript/Node.js)  ║");
  console.log("╚════════════════════════════════════════════════════════╝\n");

  // Get API key from environment
  const apiKey = process.env.PM4PY_API_KEY || "demo-key-for-testing";
  if (apiKey === "demo-key-for-testing") {
    console.log("⚠️  Using demo API key. Set PM4PY_API_KEY for real API.\n");
  }

  // Initialize client
  const client = new PM4PyClient(apiKey);

  try {
    // Example 1: Health check
    console.log("1. HEALTH CHECK");
    console.log(
      "─────────────────────────────────────────────────────────\n"
    );

    const health = await client.healthCheck();
    console.log(`✓ API Status: ${health.status}`);
    console.log(`  Version: ${health.version}`);
    console.log(`  Timestamp: ${health.timestamp}\n`);

    // Example 2: Process Discovery
    console.log("\n2. PROCESS DISCOVERY");
    console.log(
      "─────────────────────────────────────────────────────────\n"
    );

    const events = createSampleEvents();
    console.log(`Created event log with ${events.length} events\n`);

    const discoveryResult = await client.discover(events, "inductive", {
      frequency_threshold: 0.1,
    });

    const model = discoveryResult.model || {};
    console.log(`✓ Model discovered:`);
    console.log(`  Algorithm: ${discoveryResult.algorithm}`);
    console.log(`  Places: ${discoveryResult.num_places}`);
    console.log(`  Transitions: ${discoveryResult.num_transitions}`);
    console.log(`  Processing time: ${discoveryResult.processing_time_ms}ms`);
    console.log(
      `  Rate limit remaining: ${client.rateLimitRemaining} requests/hour\n`
    );

    // Example 3: Conformance Checking
    console.log("\n3. CONFORMANCE CHECKING");
    console.log(
      "─────────────────────────────────────────────────────────\n"
    );

    if (Object.keys(model).length > 0) {
      const conformanceResult = await client.conform(
        events,
        model,
        "token_replay"
      );

      const result = conformanceResult.result || {};
      console.log(`✓ Conformance checked:`);
      console.log(`  Fitness: ${formatPercent(result.fitness || 0)}`);
      console.log(`  Precision: ${formatPercent(result.precision || 0)}`);
      console.log(`  Generalization: ${formatPercent(result.generalization || 0)}`);
      console.log(`  Simplicity: ${formatPercent(result.simplicity || 0)}`);
      console.log(`  Deviant traces: ${(result.deviant_traces || []).length}`);
      console.log(`  Processing time: ${conformanceResult.processing_time_ms}ms\n`);
    }

    // Example 4: Log Statistics
    console.log("\n4. LOG STATISTICS");
    console.log(
      "─────────────────────────────────────────────────────────\n"
    );

    const statsResult = await client.stats(events);

    const stats = statsResult.stats || {};
    console.log(`✓ Statistics extracted:`);
    console.log(`  Traces: ${stats.num_traces}`);
    console.log(`  Events: ${stats.num_events}`);
    console.log(`  Activities: ${stats.num_activities}`);
    console.log(`  Mean trace length: ${(stats.trace_length_mean || 0).toFixed(2)}`);
    console.log(`  Mean case duration: ${stats.case_duration_mean}`);
    console.log(`  Processing time: ${statsResult.processing_time_ms}ms\n`);

    // Show top activities
    const activities = stats.activities || [];
    if (activities.length > 0) {
      console.log("  Top activities:");
      activities.slice(0, 3).forEach((activity) => {
        console.log(`    - ${activity.name}: ${activity.frequency} times`);
      });
    }

    // Example 5: Model Analysis
    if (Object.keys(model).length > 0) {
      console.log("\n5. MODEL ANALYSIS");
      console.log(
        "─────────────────────────────────────────────────────────\n"
      );

      const analysisResult = await client.analyze(model);

      console.log(`✓ Model analyzed:`);
      console.log(`  Is sound: ${analysisResult.is_sound}`);

      const profile = analysisResult.behavioral_profile || {};
      console.log(
        `  Strongly connected components: ${profile.strongly_connected_components}`
      );
      console.log(`  Longest path: ${profile.longest_path}`);

      const deadlocks = analysisResult.deadlock_potential || [];
      if (deadlocks.length > 0) {
        console.log(`  ⚠️  Potential deadlock configurations: ${deadlocks.length}`);
      } else {
        console.log(`  ✓ No deadlock potential`);
      }

      console.log(
        `  Processing time: ${analysisResult.processing_time_ms}ms\n`
      );
    }

    // Summary
    console.log(
      "\n╔════════════════════════════════════════════════════════╗"
    );
    console.log(
      "║                Examples Complete                       ║"
    );
    console.log(
      "╚════════════════════════════════════════════════════════╝\n"
    );

    console.log(
      `Rate limit status: ${client.rateLimitRemaining}/10000 remaining`
    );
    if (client.rateLimitReset) {
      console.log(`Rate limit resets: ${client.rateLimitReset}`);
    }
  } catch (error) {
    console.error(`\n✗ Error during examples: ${error.message}`);
    process.exit(1);
  }
}

// Run examples
main().catch(console.error);

File: examples/http_client.py
#!/usr/bin/env python3
"""
PM4Py REST API Client Examples (Python)

Demonstrates how to use the PM4Py REST API from Python applications.
This example covers:
- Authentication
- Process discovery
- Conformance checking
- Log statistics
- Error handling

Installation:
    pip install requests python-dateutil

Usage:
    export PM4PY_API_KEY="your-api-key"
    python examples/http_client.py
"""

import json
import os
import sys
from datetime import datetime, timedelta
from typing import Any, Dict, List, Optional

import requests


class PM4PyClient:
    """PM4Py REST API client with rate limit handling."""

    def __init__(self, api_key: str, base_url: str = "http://localhost:8080/api/v1"):
        """
        Initialize client.

        Args:
            api_key: API key for authentication
            base_url: Base URL of PM4Py API server
        """
        self.api_key = api_key
        self.base_url = base_url.rstrip("/")
        self.session = requests.Session()
        self.session.headers.update({"X-API-Key": api_key})
        self.rate_limit_remaining = 10000
        self.rate_limit_reset = None

    def _update_rate_limit_info(self, response: requests.Response) -> None:
        """Update rate limit information from response headers."""
        if "X-RateLimit-Remaining" in response.headers:
            self.rate_limit_remaining = int(response.headers["X-RateLimit-Remaining"])
        if "X-RateLimit-Reset" in response.headers:
            self.rate_limit_reset = response.headers["X-RateLimit-Reset"]

    def _make_request(self, method: str, path: str, data: Optional[Dict] = None) -> Dict[str, Any]:
        """
        Make HTTP request to API.

        Args:
            method: HTTP method (GET, POST, etc.)
            path: API path (without base URL)
            data: Request body data

        Returns:
            Response JSON

        Raises:
            requests.HTTPError: On HTTP errors
            ValueError: On JSON parsing errors
        """
        url = f"{self.base_url}{path}"

        try:
            if method == "GET":
                response = self.session.get(url)
            elif method == "POST":
                response = self.session.post(url, json=data)
            else:
                raise ValueError(f"Unsupported method: {method}")

            self._update_rate_limit_info(response)

            if response.status_code == 429:
                raise Exception(
                    f"Rate limit exceeded. Reset at {self.rate_limit_reset}"
                )

            response.raise_for_status()
            return response.json()

        except requests.exceptions.RequestException as e:
            print(f"HTTP Error: {e}")
            if hasattr(e.response, "text"):
                try:
                    error = e.response.json()
                    print(f"Error details: {error.get('message', 'Unknown error')}")
                except Exception:
                    print(f"Response: {e.response.text}")
            raise

    def health_check(self) -> Dict[str, Any]:
        """Check API health."""
        return self._make_request("GET", "/health")

    def discover(
        self,
        events: List[Dict],
        algorithm: str = "inductive",
        parameters: Optional[Dict] = None,
    ) -> Dict[str, Any]:
        """
        Discover process model from event log.

        Args:
            events: List of events (case_id, activity, timestamp, ...)
            algorithm: Discovery algorithm (alpha, inductive, heuristic, dfg, etc.)
            parameters: Algorithm-specific parameters

        Returns:
            Discovered model
        """
        payload = {
            "log": {
                "events": events,
                "format": "json",
            },
            "algorithm": algorithm,
        }
        if parameters:
            payload["parameters"] = parameters

        return self._make_request("POST", "/discover", payload)

    def conform(
        self,
        events: List[Dict],
        model: Dict[str, Any],
        variant: str = "token_replay",
    ) -> Dict[str, Any]:
        """
        Check log conformance against model.

        Args:
            events: List of events
            model: Petri Net model (places, transitions, arcs, etc.)
            variant: Conformance variant (token_replay, alignment)

        Returns:
            Conformance result
        """
        payload = {
            "log": {
                "events": events,
                "format": "json",
            },
            "model": model,
            "variant": variant,
        }

        return self._make_request("POST", "/conform", payload)

    def analyze(self, model: Dict[str, Any]) -> Dict[str, Any]:
        """
        Analyze model structure.

        Args:
            model: Petri Net model

        Returns:
            Analysis result
        """
        payload = {"model": model}
        return self._make_request("POST", "/analyze", payload)

    def stats(self, events: List[Dict]) -> Dict[str, Any]:
        """
        Extract log statistics.

        Args:
            events: List of events

        Returns:
            Log statistics
        """
        payload = {
            "log": {
                "events": events,
                "format": "json",
            }
        }

        return self._make_request("POST", "/stats", payload)


# ============================================================================
# EXAMPLE USAGE
# ============================================================================


def create_sample_events() -> List[Dict]:
    """Create sample event log for loan approval process."""
    events = []
    now = datetime.utcnow()

    # Create 5 sample cases
    for case_num in range(1, 6):
        case_id = f"loan_{case_num:03d}"
        case_time = now + timedelta(hours=case_num)

        # Event 1: Application
        events.append(
            {
                "case_id": case_id,
                "activity": "apply",
                "timestamp": case_time.isoformat() + "Z",
                "resource": "customer",
                "attributes": {"amount": 50000 * case_num},
            }
        )

        # Event 2: Registration
        events.append(
            {
                "case_id": case_id,
                "activity": "register",
                "timestamp": (case_time + timedelta(minutes=5)).isoformat() + "Z",
                "resource": "clerk",
            }
        )

        # Event 3: Document verification
        events.append(
            {
                "case_id": case_id,
                "activity": "verify_documents",
                "timestamp": (case_time + timedelta(hours=1)).isoformat() + "Z",
                "resource": "officer",
            }
        )

        # Event 4: Credit check
        events.append(
            {
                "case_id": case_id,
                "activity": "credit_check",
                "timestamp": (case_time + timedelta(hours=2)).isoformat() + "Z",
                "resource": "system",
            }
        )

        # Event 5: Approval
        events.append(
            {
                "case_id": case_id,
                "activity": "approve",
                "timestamp": (case_time + timedelta(hours=3)).isoformat() + "Z",
                "resource": "manager",
                "attributes": {"approved": True},
            }
        )

        # Event 6: Disbursement
        events.append(
            {
                "case_id": case_id,
                "activity": "disburse",
                "timestamp": (case_time + timedelta(hours=4)).isoformat() + "Z",
                "resource": "accountant",
                "attributes": {"amount_disbursed": 50000 * case_num},
            }
        )

    return events


def main():
    """Run example workflows."""
    print("╔════════════════════════════════════════════════════════╗")
    print("║     PM4Py REST API Client Examples (Python)            ║")
    print("╚════════════════════════════════════════════════════════╝\n")

    # Get API key from environment
    api_key = os.getenv("PM4PY_API_KEY", "demo-key-for-testing")
    if api_key == "demo-key-for-testing":
        print("⚠️  Using demo API key. Set PM4PY_API_KEY for real API.\n")

    # Initialize client
    client = PM4PyClient(api_key)

    # Example 1: Health check
    print("1. HEALTH CHECK")
    print("─────────────────────────────────────────────────────────\n")
    try:
        health = client.health_check()
        print(f"✓ API Status: {health.get('status')}")
        print(f"  Version: {health.get('version')}")
        print(f"  Timestamp: {health.get('timestamp')}\n")
    except Exception as e:
        print(f"✗ Health check failed: {e}\n")
        return

    # Example 2: Process Discovery
    print("\n2. PROCESS DISCOVERY")
    print("─────────────────────────────────────────────────────────\n")

    try:
        events = create_sample_events()
        print(f"Created event log with {len(events)} events\n")

        discovery_result = client.discover(
            events=events,
            algorithm="inductive",
            parameters={"frequency_threshold": 0.1},
        )

        model = discovery_result.get("model", {})
        print(f"✓ Model discovered:")
        print(f"  Algorithm: {discovery_result.get('algorithm')}")
        print(f"  Places: {discovery_result.get('num_places')}")
        print(f"  Transitions: {discovery_result.get('num_transitions')}")
        print(f"  Processing time: {discovery_result.get('processing_time_ms')}ms")
        print(
            f"  Rate limit remaining: {client.rate_limit_remaining} requests/hour\n"
        )

        # Example 3: Conformance Checking
        print("\n3. CONFORMANCE CHECKING")
        print("─────────────────────────────────────────────────────────\n")

        if model:
            conformance_result = client.conform(
                events=events,
                model=model,
                variant="token_replay",
            )

            result = conformance_result.get("result", {})
            print(f"✓ Conformance checked:")
            print(f"  Fitness: {result.get('fitness', 0):.2%}")
            print(f"  Precision: {result.get('precision', 0):.2%}")
            print(f"  Generalization: {result.get('generalization', 0):.2%}")
            print(f"  Simplicity: {result.get('simplicity', 0):.2%}")
            print(f"  Deviant traces: {len(result.get('deviant_traces', []))}")
            print(
                f"  Processing time: {conformance_result.get('processing_time_ms')}ms\n"
            )

        # Example 4: Log Statistics
        print("\n4. LOG STATISTICS")
        print("─────────────────────────────────────────────────────────\n")

        stats_result = client.stats(events=events)

        stats = stats_result.get("stats", {})
        print(f"✓ Statistics extracted:")
        print(f"  Traces: {stats.get('num_traces')}")
        print(f"  Events: {stats.get('num_events')}")
        print(f"  Activities: {stats.get('num_activities')}")
        print(f"  Mean trace length: {stats.get('trace_length_mean', 0):.2f}")
        print(f"  Mean case duration: {stats.get('case_duration_mean')}")
        print(
            f"  Processing time: {stats_result.get('processing_time_ms')}ms\n"
        )

        # Show top activities
        activities = stats.get("activities", [])
        if activities:
            print("  Top activities:")
            for activity in activities[:3]:
                print(
                    f"    - {activity.get('name')}: {activity.get('frequency')} times"
                )

        # Example 5: Model Analysis
        if model:
            print("\n5. MODEL ANALYSIS")
            print("─────────────────────────────────────────────────────────\n")

            analysis_result = client.analyze(model=model)

            print(f"✓ Model analyzed:")
            print(f"  Is sound: {analysis_result.get('is_sound')}")

            profile = analysis_result.get("behavioral_profile", {})
            print(f"  Strongly connected components: {profile.get('strongly_connected_components')}")
            print(f"  Longest path: {profile.get('longest_path')}")

            deadlocks = analysis_result.get("deadlock_potential", [])
            if deadlocks:
                print(f"  ⚠️  Potential deadlock configurations: {len(deadlocks)}")
            else:
                print(f"  ✓ No deadlock potential")

            print(f"  Processing time: {analysis_result.get('processing_time_ms')}ms\n")

    except Exception as e:
        print(f"✗ Error during example: {e}\n")
        sys.exit(1)

    # Summary
    print("\n╔════════════════════════════════════════════════════════╗")
    print("║                Examples Complete                       ║")
    print("╚════════════════════════════════════════════════════════╝\n")

    print(f"Rate limit status: {client.rate_limit_remaining}/10000 remaining")
    if client.rate_limit_reset:
        print(f"Rate limit resets: {client.rate_limit_reset}")


if __name__ == "__main__":
    main()
